cpp/src/parquet/encoding.cc

```diff
@@ -2466,6 +2468,99 @@ class DeltaByteArrayDecoder : public DecoderImpl,
 // ----------------------------------------------------------------------
 // BYTE_STREAM_SPLIT
 
+namespace {
+template <typename T, bool has_sse2_or_higher>
+struct ByteStreamSplitDecodePath {
+  static int Decode(T* buffer, int max_values, int num_values_in_buffer,
+                    const uint8_t* data, int& num_values_decoded, int& len_remaining) {
+    throw ParquetException("This should be unreachable.");
+  }
+};
+
+template <typename T>
+struct ByteStreamSplitDecodePath<T, true> {
+  static int Decode(T* buffer, int max_values, int num_values_in_buffer,
+                    const uint8_t* data, int& num_values_decoded, int& len_remaining) {
+    constexpr size_t num_streams = sizeof(T);
+    static_assert(num_streams == 4U || num_streams == 8U, "Invalid number of streams.");
+    constexpr size_t num_streams_log2 = (num_streams == 8U ? 3U : 2U);
+
+    const int values_to_decode = std::min(num_values_decoded, max_values);
+    const int num_decoded_previously = num_values_in_buffer - num_values_decoded;
+    const int size = values_to_decode * sizeof(T);
+    const int block_size = sizeof(__m128i) * num_streams;
+    const int num_blocks = size / block_size;
+    uint8_t* output_data = reinterpret_cast<uint8_t*>(buffer);
+
+    // First handle suffix.
```

Comment:
I think it's a very small performance difference. I moved it to the beginning to capture potential implementation bugs in which the SIMD implementation overruns the data it should work with. If the scalar code is moved to the back, it would write the correct data at the end and thus no test would fail.